### ISSUE TITLE
Assigning an array to a dataclass field with `field(init=False)`?

### ISSUE THREAD

**Issue by matthieumeo** [patrick-kidger/equinox#1070]
Hello ! 

Thanks for this amazing, beautifully designed, package. It really makes my life so much easier when developing SciML apps with JAX! 

I would like to ask one question regarding the [release notes of Equinox v0.13.0](https://github.com/patrick-kidger/equinox/releases/tag/v0.13.0). One of the points mentioned indeed: 

> * Now printing a warning if assigning an array to a dataclass field with `field(init=False)`, which can lead to unexpected behaviour. (#1038)

I don't quite understand what unexpected behaviour this may result in. Could you elaborate? I do it quite often and never encountered any side effect (at least as far as I know).

My understanding of what Equinox does behind the scenes is that is registers the dataclass as a PyTree, specifying the `tree_flatten`/`tree_unflatten` methods automatically in the process. Typically, the `__init__` method is not called when defining `tree_unflatten` for a custom PyTree, instead we use `__new__` and simply populate the attributes of the object manually, which seems to allow proper handling of dataclass fields with `field(init=False)`? 

What else could the issue be? I tried looking into #1038 but the latter does not mention anything of the kind but rather issues related to `eqx.partition` and `filter_specs` across JIT boundaries. 

Thanks a lot for the clarification! :)

**Comment by patrick-kidger** [comment]
Thank you for your kind words!

So I think your understanding of how things work behind the scenes is totally spot-on, and if you already have this mental model then I think you should be safe to ignore this warning and do anything you like.

The concrete thing this is trying to avoid is something like:
```python
class Foo(eqx.Module):
    x: int
    y: int
    z: int = eqx.field(init=False)

    def __post_init__(self):
        self.z = self.x + self.y

def consumes_foo(foo: Foo):
    # use foo.z only

jax.grad(consumes_foo)(Foo(...))
```
and then being surprised that gradients are propagated only to `.z` and not propagated to `.x` and `.y`.

In particular this behaviour is unlike PyTorch, for which the equivalent would result in `.z` holding autograd references back to `.x` and `.y`.

---

If the desire is for the gradients to land on `.x` and `.y` then it would be better to move the `z = x + y` computation into the forward pass / into a method / into a property / etc (and not a field). If the desire is for the gradients to land on `.z` then it would be better to have only `.z` as an attribute, whilst consuming `x` and `y` as `__init__`-time arguments:
```python
class Foo(eqx.Module):
    z: int
    
    def __init__(self, x: int, y: int):
        self.z = x + y
```

---

I hope that helps!

**Comment by matthieumeo** [comment]
Understood, thanks a lot for the very clear explanation! This makes complete sense indeed (but to me this behavior, even if unlike PyTorch, is actually the expected one). Closing the issue as solved :) 👍🏻
